illumos-utils/src/contract.rs

```diff
@@ -0,0 +1,387 @@
+#[cfg(target_os = "illumos")]
+use libc::ctid_t;
+use slog::{Logger, debug, error};
+use std::ffi::CString;
+use std::ffi::c_char;
+use std::ffi::c_int;
+use std::ffi::c_uint;
+use std::ffi::c_void;
+use std::thread;
+use std::time::Duration;
+
+#[allow(non_camel_case_types)]
+type ct_evthdl_t = *mut c_void;
+#[allow(non_camel_case_types)]
+type ct_evid_t = u64;
+
+#[link(name = "contract")]
+unsafe extern "C" {
+    fn ct_tmpl_set_critical(fd: c_int, events: c_uint) -> c_int;
+    fn ct_tmpl_set_informative(fd: c_int, events: c_uint) -> c_int;
+    fn ct_pr_tmpl_set_fatal(fd: c_int, events: c_uint) -> c_int;
+    fn ct_pr_tmpl_set_param(fd: c_int, params: c_uint) -> c_int;
+    fn ct_dev_tmpl_set_minor(fd: c_int, minor: *const c_char) -> c_int;
+    fn ct_tmpl_activate(fd: c_int) -> c_int;
+    fn ct_tmpl_create(fd: c_int, ctid: *mut c_int) -> c_int;
+    fn ct_tmpl_clear(fd: c_int) -> c_int;
+    fn ct_ctl_abandon(fd: c_int) -> c_int;
+    fn ct_ctl_ack(fd: c_int, evid: ct_evid_t) -> c_int;
+    fn ct_event_read_critical(fd: c_int, ev: *mut ct_evthdl_t) -> c_int;
+    fn ct_event_get_type(ev: ct_evthdl_t) -> u32;
+    fn ct_event_get_ctid(ev: ct_evthdl_t) -> ctid_t;
+    fn ct_event_get_evid(ev: ct_evthdl_t) -> ct_evid_t;
+    fn ct_event_free(ev: ct_evthdl_t);
+}
+
+// Convert an error message into an ExecutionError::ContractFailure
+fn err(msg: impl ToString) -> crate::ExecutionError {
```

Comment:
I'd drop a type alias in and a `use` here for `Result` and `Error` and make `fn err` return a `Result`:

```rust
use crate::ExecutionError as Error;
type Result<T> = std::result::Result<T, crate::ExecutionError>;

//...

fn err<T>(msg: impl ToString) -> Result<T> {
    let msg = msg.to_string();
    let err = std::io::Error::last_os_error();
    Err(Error::ContractFailure { msg, err })
}
```